File: backend/shared/utils/response.ts
import { APIGatewayProxyResult } from 'aws-lambda';

/**
 * API Response format
 */
interface ApiResponse<T = any> {
  success: boolean;
  data?: T;
  error?: {
    code: string;
    message: string;
    details?: any;
  };
  message?: string;
}

/**
 * Create success response
 */
export function successResponse<T = any>(
  data: T,
  statusCode: number = 200,
  message?: string
): APIGatewayProxyResult {
  const response: ApiResponse<T> = {
    success: true,
    data,
  };

  if (message) {
    response.message = message;
  }

  return {
    statusCode,
    headers: {
      'Content-Type': 'application/json',
      'Access-Control-Allow-Origin': process.env.CORS_ORIGINS?.split(',')[0] || '*',
      'Access-Control-Allow-Credentials': 'true',
    },
    body: JSON.stringify(response),
  };
}

/**
 * Create error response
 */
export function errorResponse(
  code: string,
  message: string,
  statusCode: number = 400,
  details?: any
): APIGatewayProxyResult {
  const response: ApiResponse = {
    success: false,
    error: {
      code,
      message,
      ...(details && { details }),
    },
  };

  return {
    statusCode,
    headers: {
      'Content-Type': 'application/json',
      'Access-Control-Allow-Origin': process.env.CORS_ORIGINS?.split(',')[0] || '*',
      'Access-Control-Allow-Credentials': 'true',
    },
    body: JSON.stringify(response),
  };
}

/**
 * Common error responses
 */
export const ErrorResponses = {
  badRequest: (message: string, details?: any) =>
    errorResponse('VALIDATION_ERROR', message, 400, details),

  unauthorized: (message: string = 'Missing or invalid authentication') =>
    errorResponse('UNAUTHORIZED', message, 401),

  forbidden: (message: string = 'Insufficient permissions') =>
    errorResponse('FORBIDDEN', message, 403),

  notFound: (resource: string, id: string) =>
    errorResponse('NOT_FOUND', `${resource} with ID '${id}' not found`, 404),

  conflict: (message: string) =>
    errorResponse('CONFLICT', message, 409),

  unprocessableEntity: (message: string, details?: any) =>
    errorResponse('SCHEMA_VALIDATION_FAILED', message, 422, details),

  rateLimitExceeded: (retryAfter: number = 60) =>
    errorResponse(
      'RATE_LIMIT_EXCEEDED',
      `Rate limit exceeded. Please retry after ${retryAfter} seconds.`,
      429,
      { retryAfter }
    ),

  internalServerError: (message: string = 'Internal server error') =>
    errorResponse('INTERNAL_SERVER_ERROR', message, 500),

  serviceUnavailable: (message: string = 'Service temporarily unavailable') =>
    errorResponse('SERVICE_UNAVAILABLE', message, 503),
};

/**
 * Pagination response format
 */
export interface PaginatedResponse<T> {
  data: T[];
  pagination: {
    total: number;
    page: number;
    limit: number;
    totalPages: number;
    hasNextPage: boolean;
    hasPreviousPage: boolean;
  };
}

/**
 * Create paginated response
 */
export function paginatedResponse<T>(
  data: T[],
  total: number,
  page: number,
  limit: number
): PaginatedResponse<T> {
  const totalPages = Math.ceil(total / limit);

  return {
    data,
    pagination: {
      total,
      page,
      limit,
      totalPages,
      hasNextPage: page < totalPages,
      hasPreviousPage: page > 1,
    },
  };
}

/**
 * CORS preflight response
 */
export function corsPreflightResponse(): APIGatewayProxyResult {
  return {
    statusCode: 200,
    headers: {
      'Access-Control-Allow-Origin': process.env.CORS_ORIGINS?.split(',')[0] || '*',
      'Access-Control-Allow-Methods': 'GET, POST, PUT, PATCH, DELETE, OPTIONS',
      'Access-Control-Allow-Headers': 'Content-Type, Authorization, X-API-Key, X-Webhook-Signature',
      'Access-Control-Allow-Credentials': 'true',
      'Access-Control-Max-Age': '86400',
    },
    body: '',
  };
}

File: backendpy/shared/utils/response.py
import json

def success_response(data, status_code=200, message=None):
    """Return success response in Lambda format"""
    body = {
        'success': True,
        'data': data
    }
    if message:
        body['message'] = message

    return {
        'statusCode': status_code,
        'headers': {
            'Content-Type': 'application/json',
            'Access-Control-Allow-Origin': '*'
        },
        'body': json.dumps(body)
    }

def error_response(message, status_code=400, code='BAD_REQUEST', errors=None):
    """Return error response in Lambda format"""
    response = {
        'success': False,
        'error': {
            'code': code,
            'message': message
        }
    }
    if errors:
        response['error']['errors'] = errors

    return {
        'statusCode': status_code,
        'headers': {
            'Content-Type': 'application/json',
            'Access-Control-Allow-Origin': '*'
        },
        'body': json.dumps(response)
    }

def bad_request(message, errors=None):
    """Return 400 Bad Request"""
    return error_response(message, 400, 'BAD_REQUEST', errors)

def unauthorized(message='Unauthorized'):
    """Return 401 Unauthorized"""
    return error_response(message, 401, 'UNAUTHORIZED')

def forbidden(message='Forbidden'):
    """Return 403 Forbidden"""
    return error_response(message, 403, 'FORBIDDEN')

def not_found(message='Resource not found'):
    """Return 404 Not Found"""
    return error_response(message, 404, 'NOT_FOUND')

def internal_server_error(message='Internal server error'):
    """Return 500 Internal Server Error"""
    return error_response(message, 500, 'INTERNAL_SERVER_ERROR')

def cors_preflight_response():
    """Return CORS preflight response in Lambda format"""
    return {
        'statusCode': 200,
        'headers': {
            'Content-Type': 'application/json',
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'GET, POST, PUT, PATCH, DELETE, OPTIONS',
            'Access-Control-Allow-Headers': 'Content-Type, Authorization, X-API-Key'
        },
        'body': ''
    }

class ErrorResponses:
    """Error response helpers"""

    @staticmethod
    def bad_request(message, errors=None):
        return error_response(message, 400, 'BAD_REQUEST', errors)

    @staticmethod
    def unauthorized(message='Unauthorized'):
        return error_response(message, 401, 'UNAUTHORIZED')

    @staticmethod
    def forbidden(message='Forbidden'):
        return error_response(message, 403, 'FORBIDDEN')

    @staticmethod
    def not_found(message='Resource not found'):
        return error_response(message, 404, 'NOT_FOUND')

    @staticmethod
    def conflict(message):
        return error_response(message, 409, 'CONFLICT')

    @staticmethod
    def internal_server_error(message='Internal server error'):
        return error_response(message, 500, 'INTERNAL_SERVER_ERROR')

def paginated_response(data, total, page, limit):
    """
    Create paginated response structure

    Args:
        data: Array of items to return
        total: Total count of items
        page: Current page number
        limit: Items per page

    Returns:
        Dict with data and pagination metadata
    """
    import math
    total_pages = math.ceil(total / limit) if limit > 0 else 0

    return {
        'data': data,
        'pagination': {
            'total': total,
            'page': page,
            'limit': limit,
            'totalPages': total_pages,
            'hasNextPage': page < total_pages,
            'hasPreviousPage': page > 1
        }
    }
